tests/returntypes/test_jvp.py

```diff
@@ -0,0 +1,719 @@
+
+    def test_jacobian_is_none_single(self):
+        """A None Jacobian returns a None JVP"""
+
+        tangent = np.array([[1.0, 2.0], [3.0, 4.0]])
+        jac = None
+
+        jvp = qml.gradients.compute_jvp_single(tangent, jac)
+        assert jvp is None
+
+    def test_jacobian_is_none_multi(self):
+        """A None Jacobian returns a None JVP"""
+
+        tangent = np.array([[1.0, 2.0], [3.0, 4.0]])
+        jac = None
+
+        jvp = qml.gradients.compute_jvp_multi(tangent, jac)
+        assert jvp is None
+
+    def test_zero_tangent_single_measurement_single_params(self):
+        """A zero dy vector will return a zero matrix"""
+        tangent = np.zeros([1])
+        jac = np.array(0.1)
+
+        jvp = qml.gradients.compute_jvp_single(tangent, jac)
+        assert np.all(jvp == np.zeros([1]))
+
+    def test_zero_tangent_single_measurement_multi_params(self):
+        """A zero tangent vector will return a zero matrix"""
+        tangent = np.zeros([2])
+        jac = tuple([np.array(0.1), np.array(0.2)])
+
+        jvp = qml.gradients.compute_jvp_single(tangent, jac)
+        assert np.all(jvp == np.zeros([1]))
+
+    def test_zero_dy_multi(self):
+        """A zero tangent vector will return a zero matrix"""
+        tangent = np.array([0.0, 0.0, 0.0])
+        jac = tuple(
+            [
+                tuple([np.array(0.1), np.array(0.1), np.array(0.1)]),
+                tuple([np.array([0.1, 0.2]), np.array([0.1, 0.2]), np.array([0.1, 0.2])]),
+            ]
+        )
+
+        jvp = qml.gradients.compute_jvp_multi(tangent, jac)
+
+        assert isinstance(jvp, tuple)
+        assert np.all(jvp[0] == np.zeros([1]))
+        assert np.all(jvp[1] == np.zeros([2]))
+
+    @pytest.mark.jax
+    @pytest.mark.parametrize("dtype1,dtype2", [("float32", "float64"), ("float64", "float32")])
+    def test_dtype_jax(self, dtype1, dtype2):
+        """Test that using the JAX interface the dtype of the result is
+        determined by the dtype of the dy."""
+        import jax
+        from jax.config import config
+
+        config.update("jax_enable_x64", True)
+
+        dtype = dtype1
+        dtype1 = getattr(jax.numpy, dtype1)
+        dtype2 = getattr(jax.numpy, dtype2)
+
+        tangent = jax.numpy.array([1], dtype=dtype1)
+        jac = tuple([jax.numpy.array(1, dtype=dtype2), jax.numpy.array([1, 1], dtype=dtype2)])
+        assert qml.gradients.compute_jvp_multi(tangent, jac)[0].dtype == dtype
+
+
+class TestJVP:
+    """Tests for the vjp function"""
+
+    def test_no_trainable_parameters(self):
+        """A tape with no trainable parameters will simply return None"""
+
+        with qml.tape.QuantumTape() as tape:
+            qml.RX(0.4, wires=0)
+            qml.CNOT(wires=[0, 1])
+            qml.expval(qml.PauliZ(0))
+
+        tape.trainable_params = {}
+        tangent = np.array([1.0])
+        tapes, fn = qml.gradients.jvp(tape, tangent, param_shift)
+
+        assert not tapes
```

Comment:
For all these tests where `tapes` is expected to be empty, it is preferable to do `assert tapes == []` instead of `assert not tapes`, since the latter would also pass if `tapes = 0` or `tapes = None`.